Adapter adc detection

yangtechen opened the pull request:
Add a vendor command 39 to read adapter ADC value.
- For the latest adapter, it will read ~48k
- For the legacy GQ/NQ/Spencer adapter, it will read ~55k

Commit:
Adaptor board detection

ADC3 will reads ~0 without adaptors and ~48000 with adapters.
Use 24000 to decide which is the case.

Commit:
make adc reading to another task

Commit:
Turn on RTX Round-Robin and add another vendor command for ADC

Commit:
rename to be compatible with upstream

gaborcsapo commented on source/rtos2/Include/RTX_Config.h:
```diff
@@ -57,7 +57,7 @@
 //   <i> Enables Round-Robin Thread switching.
 #ifndef OS_ROBIN_ENABLE
-#define OS_ROBIN_ENABLE             0 // DAPLINK. Default was: 1
+#define OS_ROBIN_ENABLE             1 // DAPLINK. Default was: 1
```
We don't want to enable this for all other HICs. Can we just set the macro in the stm32h743ii_if yaml?

gaborcsapo commented on source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_adc.h:
```diff
@@ -0,0 +1,13 @@
+#ifndef _ADAPTER_ADC_H_INCLUDED_
```
I think we need to separate the adc and the adapter_detector further.

The ADC should be initialized as part of each ADC read, because we might use the adc for other purposes in between. The adc doesn't strictly belong to the adapter detector. Epidote has something similar, you can take a look at nlnq_accessory.c

At nest, we also like to prefix our global functions with the file name. So if this file becomes adc_detector, the function names would be like adc_detector_update()

aaronokano replied:
+1 on naming

yangtechen replied:
Ok. There is an adc interface file `source/hic_hal/adc.h` after merging the upstream so I will move the logic of adc to `source/hic_hal/adc.c` and only keep adapter detector under the `extended_features/`

yangtechen replied:
`source/hic_hal/adc.h` provides an interface `adc_read_channel(uint32_t channelGroup, uint32_t channelNumber, uint32_t channelMux)`.
But as far as I know, stm32h7x does not have channel group for ADC.

aaronokano replied:
I think that's not a problem. If there's no channel group supported by this chip, then I think it should be no problem to have your implementation ignore the channelGroup parameter.

gaborcsapo commented on source/hic_hal/stm32/stm32h743ii/extended_features/include/aux_task.h:
```diff
@@ -0,0 +1,14 @@
+
+#define AUX_TASK_STACK (800)
+#define AUX_TASK_PRIORITY (osPriorityNormal)
+#define FLAGS_AUX_100MS         (1 << 0)
```
nit (small thing): the space after the macro's name is inconsistent. Generally the all macro values should be vertically aligned by using tabs.

aaronokano replied:
In addition, maybe change the name to AUX_TASK_FLAG_100MS, to be more consistent in the naming.

gaborcsapo commented on source/board/stm32h743ii.c:
```diff
@@ -23,10 +23,12 @@
 static void prerun_board_config(void) {
     I2C_DAP_Initialize();
     read_udb_version();
+    create_aux_task();
```
a more descriptive name would be udb_extended_features_task, just an idea

gaborcsapo replied:
how about adding a flag here so that when we contribute back the code, people can turn off the extended features easily.

gaborcsapo commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c:
```diff
@@ -0,0 +1,101 @@
+{
+    while (1)
+    {
+    }
```
since your next task is looked at asserts and crash info, I think asserting on errors might be more informative.

aaronokano replied:
Related question: do you have a watchdog? If you were to fall into this loop, would the board reset eventually?

yangtechen replied:
I would add a TODO comment for now. I will add crash info after I finish the logging.

yangtechen replied:
Not sure if we have a watchdog. I know that we do not turn on ST's WDG HAL driver.

gaborcsapo commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c:
```diff
@@ -0,0 +1,64 @@
+    {
+        flags = osThreadFlagsWait(FLAGS_AUX_100MS, osFlagsWaitAny, osWaitForever);
+
+        if (flags & FLAGS_AUX_100MS)
```
100ms sounds unnecessarily frequent. I think every 3 seconds is perfectly fine.

gaborcsapo commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c:
```diff
@@ -0,0 +1,64 @@
+#include "util.h"
+#include "adapter_adc.h"
+
+#ifndef USE_LEGACY_CMSIS_RTOS
```
where is this flag from? if noone would build UDB with this legacy cmsis os, then we don't need these flags and the file is more readable.

gaborcsapo commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c:
```diff
@@ -0,0 +1,64 @@
+#else
+    osTimerId_t tmr_id = osTimerNew(timer_task_100mS, osTimerPeriodic, NULL, &k_timer_100ms_attr);
+#endif
+    osTimerStart(tmr_id, 3);
```
we could move all our init code from the prerun_board_config here. As the list of features grow, it's probably better to keep it here than in source/board

gaborcsapo replied:
it's also nice to keep all the extended feature related code in one folder instead of letting it scatter out.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/include/DAP_vendor_ex.h:
```diff
@@ -10,9 +10,10 @@ typedef enum
     ID_DAP_VendorEx37_HOLD_IN_BL,
     ID_DAP_VendorEx38_RESET_DAPLINK,
+    ID_DAP_VendorEx39_READ_FLEX_HOTPLUG_ADC,
 
```
The "FLEX_HOTPLUG" seems a little strange. How about something more descriptive of what it is, like "READ_UDC_ADAPTER_TYPE_ADC"?

I'm also wondering - is it more helpful to return an ADC value or to return the type of the connected adapter? E.g., if   ~48k is type "1" and ~55k is type "2", then maybe you could have this be READ_UDC_ADAPTER_TYPE and return 0 for nothing connected, 1 for Joker board, 2 for Spencer, etc.

I'd imagine for debugging, it might be nice to have a command for returning the ADC value, but for real use seems to me like a type might be better.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/include/aux_task.h:
```diff
@@ -0,0 +1,14 @@
+#define AUX_TASK_PRIORITY (osPriorityNormal)
+#define FLAGS_AUX_100MS         (1 << 0)
+
+void create_aux_task(void);
```
Would prefer the name aux_task_create for this function

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/include/aux_task.h:
```diff
@@ -0,0 +1,14 @@
+
+void create_aux_task(void);
+
+void aux_task(void *arg);
```
Does this need to be in the header? I think it can be a static function, local to aux_task.c.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/DAP_vendor_ex.c:
```diff
@@ -271,6 +272,18 @@ uint32_t DAP_ProcessVendorCommandEx(const uint8_t *request, uint8_t *response) {
+        *response = DAP_OK;
+        *response++ = 2;
+        *response++ = ((flex_hotplug_adc_value >> 8) & 0xff);
+        *response++ = (flex_hotplug_adc_value & 0xff);
```
Out of curiosity, do you know that the response buffer has this many bytes available to write into without overrun?

yangtechen replied:
The response buffer is defined by the macro DAP_PACKET_SIZE, which is 64 bytes.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/DAP_vendor_ex.c:
```diff
@@ -271,6 +272,18 @@ uint32_t DAP_ProcessVendorCommandEx(const uint8_t *request, uint8_t *response) {
+        *response++ = ((flex_hotplug_adc_value >> 8) & 0xff);
+        *response++ = (flex_hotplug_adc_value & 0xff);
+
+        num += 3; 
```
It seems you put 4 items here. Is this number correct?

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/DAP_vendor_ex.c:
```diff
@@ -271,6 +272,18 @@ uint32_t DAP_ProcessVendorCommandEx(const uint8_t *request, uint8_t *response) {
+    case ID_DAP_VendorEx39_READ_FLEX_HOTPLUG_ADC:
+    {
+        uint16_t flex_hotplug_adc_value = get_flex_hotplug_adc_value();
+        *response = DAP_OK;
```
I think you want *response++ here, since otherwise your 2 on the next line will overwrite the DAP_OK (the post-increment will happen only after the assignment, so these first two lines will write the same address twice).

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c:
```diff
@@ -0,0 +1,101 @@
+#include "adapter_adc.h"
```
Can you add a comment at the top of this file to describe what it's for?

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c:
```diff
@@ -0,0 +1,101 @@
+
+    if (HAL_ADC_Init(&s_adc_handle) != HAL_OK)
+    {
+      error_handler();
```
nit: inconsistent indentation

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c:
```diff
@@ -0,0 +1,101 @@
+    s_adc_handle.Instance = ADC3;
+    if (HAL_ADC_DeInit(&s_adc_handle) != HAL_OK)
+    {
+      error_handler();
```
nit: inconsistent indentation

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c:
```diff
@@ -0,0 +1,101 @@
+        error_handler();
+    }
+
+    if (HAL_ADCEx_Calibration_Start(&s_adc_handle, ADC_CALIB_OFFSET, ADC_SINGLE_ENDED) != HAL_OK) {
```
nit: inconsistent placement of braces

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c:
```diff
@@ -0,0 +1,101 @@
+        error_handler();
+    }
+
+    if (HAL_ADC_PollForConversion(&s_adc_handle, 10) != HAL_OK)
```
What does the 10 mean? Is this somewhere that would be suitable to add a constant for this value?

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c:
```diff
@@ -0,0 +1,101 @@
+    }
+}
+
+void update_adc_value(void)
```
Will this only ever be called from one task? Do we need to worry about locking?

yangtechen replied:
This is only called in one task now. But yes, I think I should add a lock here to prevent the potential adc operation from  the main task.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c:
```diff
@@ -0,0 +1,101 @@
+
+uint16_t get_flex_hotplug_adc_value(void)
+{
+  return s_flex_hotplug_adc_value;
```
nit: indentation. Here, and in the next function.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c:
```diff
@@ -0,0 +1,101 @@
+void HAL_ADC_MspInit(ADC_HandleTypeDef *hadc)
+{
+  __HAL_RCC_ADC3_CLK_ENABLE();
+  UNUSED(hadc);
```
It would be more typical to put the UNUSED(hadc) at the beginning of the function rather than at the end.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c:
```diff
@@ -0,0 +1,64 @@
+#include "aux_task.h"
```
Can you add a comment to describe what this file is for? I'm not sure what an "aux task" is exactly, so it would be helpful to have an explanation.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c:
```diff
@@ -0,0 +1,64 @@
+#else
+    osTimerId_t tmr_id = osTimerNew(timer_task_100mS, osTimerPeriodic, NULL, &k_timer_100ms_attr);
+#endif
+    osTimerStart(tmr_id, 3);
```
What does the 3 mean here? Can you define some constant to give a name for it?

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c:
```diff
@@ -0,0 +1,64 @@
+void create_aux_task(void)
+{
+    adc_dap_initialize();
+#ifndef USE_LEGACY_CMSIS_RTOS
```
Based on Gabor's comment, I'm guessing these ifdefs will get deleted, but if they stick around, then please flip the logic to ifdef-else instead of ifndef-else - it's a little bit easier to understand that way.

aaronokano commented on source/rtos2/Include/RTX_Config.h:
```diff
@@ -57,7 +57,7 @@
 //   <i> Enables Round-Robin Thread switching.
 #ifndef OS_ROBIN_ENABLE
-#define OS_ROBIN_ENABLE             0 // DAPLINK. Default was: 1
+#define OS_ROBIN_ENABLE             1 // DAPLINK. Default was: 1
```
Why is this needed?

yangtechen replied:
We want to put our own logic in another thread, I need to enable Round-Robin so that the scheduler could preempt the main task.

aaronokano replied:
Does the main task never yield?

yangtechen replied:
Yes, if the we left the Round-Robin disabled, the aux_task will never run.

Commit:
Separate ADC from adapter detector

Commit:
fix style

aaronokano commented on source/hic_hal/stm32/stm32h743ii/adc.c:
```diff
@@ -0,0 +1,105 @@
+#include "stm32h7xx_hal.h"
+
+// Timeout value in millisecond
+#define ADC_CONVERSION_TIMEOUT (10)
```
Can you make this ADC_CONVERSION_TIMEOUT_MS, so the unit is clear in the location where it is used?

aaronokano commented on source/hic_hal/stm32/stm32h743ii/adc.c:
```diff
@@ -0,0 +1,105 @@
+
+uint32_t adc_read_channel(uint32_t channelGroup, uint32_t channelNumber, uint32_t channelMux)
+{
+    uint16_t udc_adapter_adc_value;
```
I think just adc_value is ok. This looks like it's supposed to be somewhat generic, so not limited to reading the UDC adapter type, is that right?

aaronokano commented on source/hic_hal/stm32/stm32h743ii/adc.c:
```diff
@@ -0,0 +1,105 @@
+{
+}
+
+uint32_t adc_read_channel(uint32_t channelGroup, uint32_t channelNumber, uint32_t channelMux)
```
Should this function multiple tasks attempting an ADC conversion simultaneously? How is this done in other ADC drivers?

yangtechen replied:
I will add a mutex to prevent such race condition.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h:
```diff
@@ -0,0 +1,20 @@
+#define ADAPTER_JOKER_FLEX_ADC_MIN      (51000)
+#define ADAPTER_UDB_6_FLEX_ADC_MIN      (47000)
+#define ADAPTER_UDB_17_FLEX_ADC_MIN     (44000)
+#define ADAPTER_UDB_12_FLEX_ADC_MIN     (40000)
```
Are there already reserved values for future adapters? Can we go ahead and set them up here ahead of time so that we can return a correct adapter_type_t even if we don't know which specific adapter it is yet?

yangtechen replied:
I have no idea if there are reserved value. But I think it is hard to name those adapters, I will just use ADAPTER_UNKNOWN now.

aaronokano replied:
I think it would be fine to just write something like ADAPTER_TYPE_6, ADAPTER_TYPE_7, etc. The reason I'm suggesting this is because at some point in the future, people will be making new adapters and you probably do not want to require a UDB firmware update for each one just to read which type it is. It's better to have a mapping established ahead of time so at least the type value can be read for each one, even if you don't internally have a name for it.

yangtechen replied:
I found a pin-mapping for UDB, it seems there are some reserved values https://docs.google.com/spreadsheets/d/1crGH2GwyluorJVSYZoUDuaexZTivegYbg4MbRePD_6s/edit#gid=0

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h:
```diff
@@ -0,0 +1,20 @@
+
+typedef enum
+{
+    ADAPTER_UDB_13_FLEX,
```
Should there be an ADAPTER_NONE value? Having nothing attached and having an unknown adapter attached are two different cases.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h:
```diff
@@ -0,0 +1,20 @@
+    ADAPTER_UNKNOWN
+} adapter_type_t;
+
+#define ADAPTER_UDB_13_FLEX_ADC_MIN     (54000)
```
Would this be better in a table of some sort to allow easy iteration?

    // Table to map adapter type to min ADC value that can be read for type
    static const uint16_t s_adapter_type_adc_map[] = {
        [ADAPTER_UDB_13_FLEX] = 54000,
        [ADAPTER_UDB_JOKER_FLEX] = 51000,
        ...
    };

Also might be better to put it into the C file where it's used, since I suspect it will only be one file that needs this info.

yangtechen replied:
Good suggestion, see this syntax the first time.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/include/udb_extended_features_task.h:
```diff
@@ -0,0 +1,20 @@
+
+#include "cmsis_os2.h"
+
+#define UDB_EXTENDED_FEATURES_TASK_STACK          (800)
```
I think these defines can all be moved into udb_extended_features_task.c where they are used. No need to cause other files to include more definitions than necessary.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/udb_extended_features_task.c:
```diff
@@ -0,0 +1,55 @@
+{
+    uint16_t flags = 0;
+
+    adc_init();
```
Don't we read the HW ID before this point? Doesn't the HW ID require the ADC? We should probably read it after initializing the ADC.

yangtechen replied:
ADC is only used for adapter detector. HW ID is detected by GPIO read.

aaronokano replied:
Really? I'm not sure if I have the correct schematics, but it seems like proto2 has a voltage divider to express the version and is connected to an ADC input. I guess if we just want to distinguish proto1 and proto2, a GPIO could work (although I don't quite understand how it can work at all in proto1 regardless, since the pin is floating).

I suppose addressing the hardware version reading can be done later, but maybe it would be good to pre-emptively move the HW version read to after ADC init, since at some point, at the very minimum when the next hardware revision after proto2 comes out, it's going to need to use the ADC.

yangtechen replied:
ADC1_2_3_EXTI15_R is just for ADC external trigger and can not be used as ADC input. So what Eric came up with is that we can enable internal weak pull-up register for the GPIO to make proto1 read high and proto2 read low. But you are right that this can only differentiate p1 and p2 and can not be used for future versions.

aaronokano replied:
How is this going to work for future versions? Are they going to move to a different pin?

This sounds like it's going to be a headache for builds after p2. Anyway, I think this is a discussion for another time.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/include/udb_extended_features_task.h:
```diff
@@ -0,0 +1,20 @@
+#define UDB_EXTENDED_FEATURES_TASK_PRIORITY       (osPriorityNormal)
+#define UDB_EXTENDED_FEATURES_TASK_FLAGS_3S       (1 << 0)
+
+// 1 tick is 10ms
```
Is the 10ms tick a fixed value? If not, is there a way to retrieve the tick rate and define this value in terms of the tick rate?

yangtechen replied:
osKernelGetTickFreq() could work

Commit:
Fix style further

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_detector.h:
```diff
@@ -0,0 +1,21 @@
+
+typedef enum
+{
+    ADAPTER_UDB_13_FLEX,
```
One more suggestion on the adapter type enum - I suggest explicitly listing the value of each enum member so that one can easily look at the enum and know which number corresponds to which type.

aaronokano replied:
Sorry, also meant to add an example of what I meant:

    typedef enum
    {
        ADAPTER_UDB_13_FLEX = 0,
        ADAPTER_JOKER_FLEX  = 1,
        ...
    };

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_detector.h:
```diff
@@ -0,0 +1,21 @@
+    ADAPTER_UDB_6_FLEX,
+    ADAPTER_UDB_17_FLEX,
+    ADAPTER_UDB_12_FLEX,
+    ADAPTER_UNKNOWN,
```
I suggest reserving a number outside of the range of normal types for the UNKNOWN and NONE types. You could maybe have ADAPTER_NONE = 0 and ADAPTER_UNKNOWN = 0xFF or something along those lines (not sure what value is most suitable here). This is because at some point you will add more adapter types, and then the numbers that used to mean UNKNOWN and NONE will mean something else. Better to keep them the same, especially if these types will be part of the software interface at some point.

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_detector.c:
```diff
@@ -0,0 +1,50 @@
+#include "stm32h7xx_hal.h"
+#include "adc.h"
+
+#define ADAPTER_DETECTOR_ADC_CHANNEL ADC_CHANNEL_9
```
What pin does this correspond to? Can you add a comment for that?

Commit:
Add reserved adapter type and locking mechanism

1. Since adc.c use osMutex_t, move it to extended_features/ so that
the bootloader will not include the file during compilation.
2. Add reserved adapter resister adc value.
3. Lock adc_read_channel() to prevent race condition.

Commit:
delete debug and unused macros

Commit:
fix implict included name WORDS and add ADAPTER_UNKNOWN comment

aaronokano commented on source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_detector.c:
```diff
@@ -0,0 +1,64 @@
+    [ADAPTER_TYPE_13]       = 11040,
+    [ADAPTER_TYPE_14]       = 7567,
+    [ADAPTER_TYPE_15]       = 5904,
+    [ADAPTER_UNKNOWN]       = 0xff,
```
Can you add a comment to say that this is an arbitrarily chosen low value between 0 and adapter type 15?

aaronokano commented:
This looks good to me! Thanks, Yang-te. Looks like Gabor or Eric needs to give the final approval.
